### PR TITLE
plugins {} block supports plugin requests for plugins already present on the classpath

### PR DESCRIPTION
**Pull request by eskatos**
This allows the following use cases using the `plugins {}` block:

1. request `buildSrc` plugin

    ```kotlin
    // buildSrc/src/main/kotlin/build/MyPlugin.kt
    package build

    import org.gradle.api.*

    open class MyPlugin : Plugin<Project> {

        override fun apply(project: Project) {
            // ...
        }
    }
    ```

    ```kotlin
    // buildSrc/build.gradle.kts
    plugins {
        `kotlin-dsl`
        `java-gradle-plugin`
    }

    gradlePlugin {
        (plugins) {
            "my-plugin" {
                id = "my-plugin"
                implementationClass = "build.MyPlugin"
            }
        }
    }
    ```

    ```kotlin
    // build.gradle.kts
    plugins {
        `my-plugin`
    }
    ```

2. request plugin applied to parent project

    ```kotlin
    // root/build.gradle.kts
    plugins {
      id("foo") version "1.0"
    }

    // root/sub/build.gradle.kts
    plugins {
      id("foo")
    }
    ```

3. request plugin requested on but not applied to parent project

    ```kotlin
    // root/build.gradle.kts
    plugins {
      id("foo") version "1.0" apply false
    }

    // root/sub/build.gradle.kts
    plugins {
      id("foo")
    }
    ```

Requesting a version of a plugin already on the classpath is forbidden.

See https://github.com/gradle/kotlin-dsl/issues/426

### CODE REVIEW

**Comment by oehme** [comment]
Consider always failing when a version is specified in the child. It is redundant and any change to the parent would make the child fail anyway.

**Comment by eskatos** [comment]
Very good point @oehme.
I made the changes.

In exchange for the classloader scope meta data we need to preserve the original plugin request version when a plugin resolution rule changes it. Because in the presence of a plugin resolution rule that call `useVersion(..)` the plugin requests will have a `version` set on the whole project hierarchy. And we don’t want to fail in this case.

**Comment by oehme** [comment]
Good idea, we'll want that for reporting purposes anyway I think. Maybe store the whole original request instead of just the original version?

**Comment by eskatos** [comment]
Done

File: subprojects/plugin-use/src/integTest/groovy/org/gradle/plugin/use/AlreadyOnClasspathPluginUseIntegrationTest.groovy [hunk 1]
```diff
@@ -0,0 +1,323 @@
...
+        operations.hasOperation("Apply plugin my-plugin to project ':a'")
+    }
+
+    def "can request non-core plugin already applied to parent project"() {
+
+        given:
+        settingsFile << "include('a')\n"
+
+        and:
+        buildFile << requestPlugin("com.gradle.plugin-publish", "0.9.8")
```
Comment (bamboo) [inline subprojects/plugin-use/src/integTest/groovy/org/gradle/plugin/use/AlreadyOnClasspathPluginUseIntegrationTest.groovy]: ⭕️  My preference would be to use a plugin local to the test, like publishing a small plugin to a local ivy repository during `setup` or something.

File: subprojects/plugin-use/src/integTest/groovy/org/gradle/plugin/use/AlreadyOnClasspathPluginUseIntegrationTest.groovy [hunk 2]
```diff
@@ -0,0 +1,323 @@
...
+
+        and:
+        buildFile << requestPlugin("my-plugin", "1.0")
+
+        when:
+        fails "help"
+
+        then:
+        failureDescriptionStartsWith("Error resolving plugin [id: 'my-plugin', version: '1.0']")
+        failureHasCause("Cannot request a version for a plugin already present on the classpath")
```
Comment (bamboo) [inline subprojects/plugin-use/src/integTest/groovy/org/gradle/plugin/use/AlreadyOnClasspathPluginUseIntegrationTest.groovy]: :o: I would prefer a message that tells me what I should do to fix the problem.

**Comment by eskatos** [comment]
Comments addressed.
@oehme, @bamboo, could you have another look?

**Comment by bamboo** [comment]
👍

**Pull request merged**